### PR TITLE
New rule: prefer_conditional_assignment

### PR DESCRIPTION
**Pull request by JPaulsen**
Fixes https://github.com/dart-lang/linter/issues/483

### CODE REVIEW

File: lib/src/rules/prefer_conditional_assignment.dart [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+**GOOD:**
+```
+String get fullName {
+  return _fullName ??= getFullUserName(this);;
+}
+```
+
+''';
+
+Element _checkConditionAndGetElement(Expression node) {
```
Comment (JPaulsen) [inline lib/src/rules/prefer_conditional_assignment.dart]: I have to split this method in 2 methods

Comment (JPaulsen) [inline lib/src/rules/prefer_conditional_assignment.dart]: Fixed

**Comment by JPaulsen** [comment]
178 lints in the SDK, no false positives found.

File: lib/src/rules/prefer_conditional_assignment.dart [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  final LintRule rule;
+  _Visitor(this.rule);
+
+  @override
+  visitIfStatement(IfStatement node) {
+    final elementInCondition = _getElementInCondition(node.condition);
+    if (elementInCondition != null &&
+        _checkStatement(node.thenStatement, elementInCondition)) {
+      rule.reportLint(node);
+    }
```
Comment (bwilkerson) [inline lib/src/rules/prefer_conditional_assignment.dart]: What about
```
if (x == null) {
  x = 'string';
} else {
  x = '$x$x';
}
```

File: test/rules/prefer_conditional_assignment.dart [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+class Person {
+  int x;
+  String _fullName;
+  String get badWithBlock1 {
+    if (_fullName == null) { // LINT
+      _fullName = getFullUserName(this);
+    }
+    return _fullName;
```
Comment (bwilkerson) [inline test/rules/prefer_conditional_assignment.dart]: The `return` statements are uninteresting. Consider removing them.

File: lib/src/rules/prefer_conditional_assignment.dart (hunk 2 above)
Comment (JPaulsen) [inline lib/src/rules/prefer_conditional_assignment.dart]: I had that in mind and I forgot it!, sorry, elseStatement must be null!, thanks.

Comment (JPaulsen) [inline lib/src/rules/prefer_conditional_assignment.dart]: Fixed

File: test/rules/prefer_conditional_assignment.dart (hunk 3 above)
Comment (JPaulsen) [inline test/rules/prefer_conditional_assignment.dart]: Fixed

**Pull request merged**